File: rest-remote/src/main/java/com/jgoticks/RemoteLookupProxy.java
package com.jgoticks;

import akka.actor.AbstractActor;
import akka.actor.ActorIdentity;
import akka.actor.ActorRef;
import akka.actor.ActorSelection;
import akka.actor.Identify;
import akka.actor.Terminated;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import akka.util.Timeout;
import scala.concurrent.duration.FiniteDuration;

public class RemoteLookupProxy extends AbstractActor {
    private String path_;
    private LoggingAdapter log = Logging.getLogger(this);

    public RemoteLookupProxy(String path) {
        path_ = path;
        getContext().setReceiveTimeout(FiniteDuration.create(3, "second"));
        sendIdentifyRequest();
    }

    private void sendIdentifyRequest() {
        ActorSelection selection = getContext().actorSelection(path_);
        selection.tell(new Identify(path_), getSelf());
    }

    @Override
    public Receive createReceive() {
        return identify();
    }

    private Receive identify() {
        return receiveBuilder()
                .match(ActorIdentity.class, msg -> {
                    msg.getActorRef().ifPresent(ref -> {
                        getContext().setReceiveTimeout(FiniteDuration.create(3, "s"));
                        log.info("switching to active state");
                        getContext().become(active(ref));
                        getContext().watch(ref);
                    });
                })
                .build();
    }

    private Receive active(ActorRef actor) {
        return receiveBuilder()
                .match(Terminated.class, msg -> {
                    log.info("actor " + msg.getActor() + " terminated.");
                    getContext().become(identify());
                    log.info("switching to identify state");
                    getContext().setReceiveTimeout(FiniteDuration.create(3, "second"));
                    sendIdentifyRequest();
                })
                .matchAny(msg -> actor.forward(msg, getContext()))
                .build();
    }
}

File: rest-remote/src/main/scala/com/goticks/remote/RemoteLookupProxy.scala
package com.goticks.remote

import akka.actor.{Actor, ActorIdentity, ActorLogging, ActorRef, Identify, Terminated}

import scala.concurrent.duration._

class RemoteLookupProxy(path: String)
  extends Actor
  with ActorLogging {

  context.setReceiveTimeout(3 second)
  sendIdentifyRequest()

  def sendIdentifyRequest(): Unit = {
    log.info("send identify request=>{}", path)
    val selection = context.actorSelection(path)
    selection ! Identify(path)
  }

  def receive = identify

  def identify: Receive = {
    case ActorIdentity(path, Some(actor)) =>
      context.setReceiveTimeout(Duration.Undefined)
      log.info("switching active state")
      context.become(active(actor))
      context.watch(actor)
  }

  def active(actorRef: ActorRef): Receive = {
    case Terminated(actor) =>
      log.info(s"Actor $actor terminated.")
      log.info("switching to identify state")
      context.become(identify)
      context.setReceiveTimeout(3 second)
      sendIdentifyRequest()
    case msg: Any => actorRef forward msg
  }
}
